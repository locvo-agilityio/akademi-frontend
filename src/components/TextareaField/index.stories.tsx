import { StoryObj, Meta } from '@storybook/react';

import TextareaField from '.';

const defaultProps = {
  label: 'Email',
  name: 'email',
  placeholder: 'Email',
  onChange: () => {},
};

const meta = {
  title: 'Components/TextareaField',
  component: TextareaField,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],
} satisfies Meta<typeof TextareaField>;

export default meta;

type Story = StoryObj<typeof TextareaField>;

export const Default: Story = {
  args: {
    ...defaultProps,
  },
};

export const HasError: Story = {
  args: {
    ...defaultProps,
    isError: true,
  },
};
