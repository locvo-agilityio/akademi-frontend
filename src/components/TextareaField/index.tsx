'use client';

import { ChangeEvent, memo } from 'react';
import {
  FormLabel,
  FormErrorMessage,
  FormControl,
  Textarea,
  TextareaProps,
} from '@chakra-ui/react';

type TTextareaFieldProps = Omit<TextareaProps, 'onChange'> & {
  isError?: boolean;
  errorMessages?: string;
  label?: string;
  onChange: (value: string) => void;
};

const TextareaField = ({
  isError = false,
  errorMessages = 'Default error',
  label = '',
  onChange,
  ...rest
}: TTextareaFieldProps) => {
  const handleChangeValue = (e: ChangeEvent<HTMLTextAreaElement>): void =>
    onChange(e.target.value);

  return (
    <FormControl isInvalid={isError}>
      {label && (
        <FormLabel
          fontSize="md"
          fontWeight="semibold"
          marginInlineEnd={0}
          minW="max-content"
          color={isError ? 'red' : 'darkBlue'}
        >
          {label} *
        </FormLabel>
      )}

      <Textarea
        variant="primary"
        onChange={handleChangeValue}
        {...rest}
        isInvalid={isError}
      />
      {isError && (
        <FormErrorMessage color="red" fontSize="2xs">
          {errorMessages}
        </FormErrorMessage>
      )}
    </FormControl>
  );
};

export default memo(TextareaField);
