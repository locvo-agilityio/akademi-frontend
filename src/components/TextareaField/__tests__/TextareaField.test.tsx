import { fireEvent, render } from '@testing-library/react';

// Components
import TextareaField from '..';

describe('TextareaField component', () => {
  const mockOnChange = jest.fn();
  const props = {
    label: 'Description',
    name: 'description',
    placeholder: 'Description',
    onChange: mockOnChange,
  };
  it('render correctly with default props', () => {
    const { container } = render(<TextareaField {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('render correctly with error message', () => {
    const { container } = render(
      <TextareaField {...props} isError errorMessages="Field is required" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should call onChange when input value is changed', async () => {
    const { getByPlaceholderText } = render(<TextareaField {...props} />);
    fireEvent.change(getByPlaceholderText('Description'), {
      target: { value: 'user@gmail.com' },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });
});
