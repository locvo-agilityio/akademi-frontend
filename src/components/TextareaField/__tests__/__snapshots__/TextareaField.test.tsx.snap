// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaField component render correctly with default props 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-1hm3o85"
      for="field-:r0:"
      id="field-:r0:-label"
    >
      Description
       *
    </label>
    <textarea
      class="chakra-textarea css-0"
      id="field-:r0:"
      name="description"
      placeholder="Description"
    />
  </div>
</div>
`;

exports[`TextareaField component render correctly with error message 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    data-invalid=""
    role="group"
  >
    <label
      class="chakra-form__label css-1ooge5m"
      data-invalid=""
      for="field-:r1:"
      id="field-:r1:-label"
    >
      Description
       *
    </label>
    <textarea
      aria-describedby="field-:r1:-feedback"
      aria-invalid="true"
      class="chakra-textarea css-0"
      id="field-:r1:"
      name="description"
      placeholder="Description"
    />
    <div
      aria-live="polite"
      class="chakra-form__error-message css-15e0ffo"
      id="field-:r1:-feedback"
    >
      Field is required
    </div>
  </div>
</div>
`;
